scene/gui/button.cpp

```diff
@@ -72,287 +129,266 @@ void Button::_notification(int p_what) {
 		} break;
 
 		case NOTIFICATION_DRAW: {
-			RID ci = get_canvas_item();
-			Size2 size = get_size();
-			Color color;
-			Color color_icon(1, 1, 1, 1);
+			const RID ci = get_canvas_item();
+			const Size2 size = get_size();
 
-			Ref<StyleBox> style = theme_cache.normal;
-			bool rtl = is_layout_rtl();
-			const bool is_clipped = clip_text || overrun_behavior != TextServer::OVERRUN_NO_TRIMMING;
+			const Ref<StyleBox> style = _get_current_stylebox();
+			{ // Draws the stylebox in the current state.
+				if (!flat) {
+					style->draw(ci, Rect2(Point2(), size));
+				}
 
-			switch (get_draw_mode()) {
-				case DRAW_NORMAL: {
-					if (rtl && has_theme_stylebox(SNAME("normal_mirrored"))) {
-						style = theme_cache.normal_mirrored;
-					} else {
-						style = theme_cache.normal;
-					}
+				if (has_focus()) {
+					Ref<StyleBox> style2 = theme_cache.focus;
+					style2->draw(ci, Rect2(Point2(), size));
+				}
+			}
 
-					if (!flat) {
-						style->draw(ci, Rect2(Point2(0, 0), size));
-					}
+			Ref<Texture2D> _icon = icon;
+			if (_icon.is_null() && has_theme_icon(SNAME("icon"))) {
+				_icon = theme_cache.icon;
+			}
+
+			if (xl_text.is_empty() && _icon.is_null()) {
+				break;
+			}
+
+			const float style_margin_left = style->get_margin(SIDE_LEFT);
+			const float style_margin_right = style->get_margin(SIDE_RIGHT);
+			const float style_margin_top = style->get_margin(SIDE_TOP);
+			const float style_margin_bottom = style->get_margin(SIDE_BOTTOM);
+
+			Size2 drawable_size_remained = size;
 
+			{ // The size after the stelybox is stripped.
+				drawable_size_remained.width -= style_margin_left + style_margin_right;
+				drawable_size_remained.height -= style_margin_top + style_margin_bottom;
+			}
+
+			const int h_separation = MAX(0, theme_cache.h_separation);
+
+			// The width reserved for internal element in derived classes (and h_separation if need).
+			float internal_margin = _internal_margin[SIDE_LEFT] > _internal_margin[SIDE_RIGHT] ? _internal_margin[SIDE_LEFT] : _internal_margin[SIDE_RIGHT];
```

Comment:
I still don't understand why you're doing it this way. Why aren't we considering both left and right margin? What if I need to have margins on both sides? Currently, `OptionButton` is not set up this way, but that's still unreliable overall.

Reply:
There is no ready-made example, and it is a bit difficult to imagine its rules. In that case, `h_separation` may also need to be calculated twice?

Reply:
I guess, yes 👍